### ISSUE TITLE
KeepAliveInterval doesn't seem to work

### ISSUE THREAD

**Issue by aspnet-hello** [aspnet/AspNetCore#2464]
_From @nehmebilal on Thursday, October 5, 2017 9:54:50 AM_

In my StartUp class of ASP.Net app hosted using WebListener, I am adding the WebSocketMiddleware as follows:

```csharp
            var webSocketOptions = new WebSocketOptions
            {
                KeepAliveInterval = TimeSpan.FromSeconds(30),
            };
            app.UseWebSockets(webSocketOptions);
```

Then I am creating WebSockets as follows:

```csharp
        public async Task Configure(HttpContext hc, Func<Task> next)
        {
           var webSocket = hc.WebSockets.AcceptWebSocketAsync();
           /// ...
        }
```

The value I use for KeepAliveInterval doesn't seem to have any effect and it still seems to take up to 2 minutes to send a ping.

I am currently using Microsoft.AspNetCore.WebSockets 1.0.2.

Am I doing anything wrong?

_Copied from original issue: aspnet/WebSockets#204_

**Comment by aspnet-hello** [comment]
_From @Tratcher on Thursday, October 5, 2017 10:08:00 AM_

Takes up to 2 minutes for what? To send a Ping?

**Comment by aspnet-hello** [comment]
_From @nehmebilal on Thursday, October 5, 2017 10:08:43 AM_

Yes (description updated).

**Comment by aspnet-hello** [comment]
_From @nehmebilal on Friday, October 6, 2017 8:43:37 AM_

Any thoughts?

**Comment by aspnet-hello** [comment]
_From @Tratcher on Wednesday, October 11, 2017 10:41:53 AM_

The relevant code:
https://github.com/aspnet/WebSockets/blob/c65e8681c4a50fc33c263d1f42dfcf007cb5c035/src/Microsoft.AspNetCore.WebSockets/Internal/fx/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ManagedWebSocket.cs#L487-L501
It does have that else clause that says it will skip sending a Ping if any other send is already in progress.

**Comment by aspnet-hello** [comment]
_From @nehmebilal on Wednesday, October 11, 2017 10:51:23 AM_

Ok. In my experiments, there was pretty much no traffic other than the pings so there shouldn't be any contention on the lock.

Is it possible that the options passed through `app.UseWebSockets(options)` is not making it to the `ManagedWebSocket`?

**Comment by jekusz** [comment]
Have just been playing with the websocket basically deploying this demo implementation locally: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/websockets?view=aspnetcore-2.0
and I find that there are no ping frames sent through what-so-ever. Have actually tested it by connecting another machine and pulling its network cable -> 10 minutes later it's still registering as a connected user to the server. These are my settings:
```
var webSocketOptions = new WebSocketOptions()
        {
            KeepAliveInterval = TimeSpan.FromSeconds(1),
            ReceiveBufferSize = 4 * 1024
        };
        app.UseWebSockets(webSocketOptions);
```
And these are my versions:
Versions used - AspNetCore.All (2.0.7) -> WebSockets (2.0.2)

**Comment by Tratcher** [comment]
Which server? There are some known issues for IIS/Express.

**Comment by jekusz** [comment]
tested on:
IIS express 10.0.14358.10000
IIS 10.0.17134.1

**Comment by Tratcher** [comment]
@pan-wang keep alives aren't forwarded through IIS/ANCM, right? Is there anyway to control IIS's outbound websocket keep alives?

**Comment by pan-wang** [comment]
can you try config the ping interval of the IIS application to see whether it helps?
https://docs.microsoft.com/en-us/iis/configuration/system.webserver/websocket, e.g., adding the following part to the web.config <system.webServer>
   <webSocket
      enabled="true" 
      receiveBufferLimit="4194304"
      pingInterval="00:01:00">
   </webSocket>
</system.webServer>

**Comment by jekusz** [comment]
The applicationHost.config has by default this:
<section name="webSocket" overrideModeDefault="Deny" /> which is causing:
`
This configuration section cannot be used at this path. This happens when the section is locked at a parent level. Locking is either by default (overrideModeDefault="Deny"), or set explicitly by a location tag with overrideMode="Deny" or the legacy allowOverride="false".
--
`
however, once it's changed to 'allow', and the app starts with the altered web.config, there are still no pings being received (note: I'm now looking at frames in chrome dev-tools only )

**Comment by pan-wang** [comment]
@jekusz  browser dev-tool will not show websocket ping/pong. You need install wireshark. After filter websocket.payload.ping, you will see there are ping/pong between browser and server.

**Comment by jekusz** [comment]
@pan-wang 
On the modified web.config, -> I am getting websocket pingpong in wireshark indeed. The timing is very strange (it seems to be double to what its set in web.config). They stop after a few minutes though.
On the original web.config (i.e. without the websocket tag) -> using wireshark, i dont get any ping at any point.
I don't see any effect of the KeepAliveInterval setting on anything.

**Comment by pan-wang** [comment]
@jekusz  the communication mode is client<-->IIS/ANCM<-->Kestrel.  The ping/pong between client and IIS/ANCM is handled by IIS websocket module (ANCM never sees the ping request and thus there is no forwarding). the Ping/Pong between IIS/ANCM and kestrel is handled by WinHttp which is used by ANCM. ANCM also never sees the ping/pong message. This is by design as ping/pong is not treated as user data.

**Comment by DemoBytom** [comment]
So, is there any way to detects a broken connection from the client, when running an app behind IIS?
I have a web api that receives data through websocket. If I break the connection during the process, the application is stuck at the
```csharp
result = await websocket.ReceiveAsync(
    new ArraySegment<byte>(buffer),
    new CancellationTokenSource(500).Token);
```
The CancellationToken doesn't cancel the call, and `KeepAlive` mechanic doesn't kick in either, I'm just stuck and can't react to broken connection (which I need to do).

If I run an app on Kestrel alone `ReceiveAsync` throws a `System.Net.WebSockets.WebSocketException` and I can react to that.

**Comment by andriusbutk** [comment]
Hey, what if I have implemented websocket and the client is the one who is willing to send Ping, and my server has to response with Pong? Is it somehow possible using Websockets in .net core?
And in my situation the websocket client MUST send Ping by the way.

**Comment by anurse** [comment]
It's not possible to configure the ping/pong frames between IIS and Kestrel when using the out-of-proc version of ANCM. Ping/pongs are entirely controlled by IIS and aren't visible to the application. If you need an application-specific ping/pong behavior, you'll have to put that in your own logic.

**Comment by nehmebilal** [comment]
@anurse Can you explain then what the `KeepAliveInterval` parameter is for? If it has no effect, shouldn't be removed?

**Comment by anurse** [comment]
The KeepAliveInterval **does** control how frequently the server sends Ping messages, the problem is that when running behind IIS, those messages are received by IIS and not forwarded to the client. It should still work fine when using Kestrel directly.

This could also be true of any downstream proxy. We don't know about those, so we have the API to control our behavior.

**Comment by nehme-bilal-ah** [comment]
That’s good to know!

**Comment by onetsmr** [comment]
@aspnet-hello 

So KeepAliveInterval still does not work as expected on Kestrel:
1. I set KeepAliveInterval=10 seconds on the server
2. Then I set break point in the client code 
3. Then run and I see "ping" messages from server every 10 sec. Client sends "pong" message.
4. Then I terminate internet connection on the client, but the server does not know that the client has fallen off either after 10 seconds or after a minute.

I see in the documentation https://docs.microsoft.com/en-us/aspnet/core/fundamentals/websockets?view=aspnetcore-3.0#handle-client-disconnects that I should send ping/pong. But KeepAlive is the ping/pong. So server should send "ping", wait "KeepAliveInterval" x2 seconds and if it doesn't recive "pong" from the client - terminate connection and set aborted (?) state for the socket.

It doesn't work like that. No exception, no aborted state, nothing...
